docs/adrs/024-remote-module-invocation.md

```diff
@@ -0,0 +1,280 @@
+        if self.connection.protocol == "grpc":
+            <run remote `method` using grpc>
+        elif self.connection.protocol == "http":
+            <run remote `method` using http>
+```
+
+## RemoteModuleConfig
+
+```yaml
+class RemoteModuleConfig(ModuleConfig):
+  # Remote runtime information copied from the RemoteModelFinder config
+  connection: Dict[str, Any]
+
+  # Method information
+  # use list and tuples instead of a dictionary to avoid aconfig.Config error
+  task_methods: List[Tuple[type[TaskBase], List[RemoteMethodRpc]]]
+  train_method: RemoteMethodRpc
+  
+  # Source Module Information
+  module_id: str
+  module_name: str
+  model_path: str
+
+@dataclass
+class RemoteMethodRpc:
+    # full signature for this RPC
+    signature: CaikitMethodSignature
+    
+    # Request and response objects for this RPC
+    request_dm_name: str
+    response_dm_name: str
+    
+    # Either the function name of the GRPC Servicer or HTTP endpoint
+    rpc_name: str
+
+    # Only used for infer RPC types
+    input_streaming: bool
+    output_streaming: bool
+```
+
+The RemoteModuleConfig is a custom subclass of ModuleConfig, which contains a description of a Module's
+tasks, inference and train methods, and version information, as well as the connection information of 
+the remote runtime. The combination of the two allows the RemoteModelInitializer to construct a new 
+RemoteModule without having to import `caikit.runtime` or the source model. To help simplify the
+config definition and access, we decided to create a helper dataclass, `RemoteMethodRpc`, which contains
+information about a specific method and includes things like the CaikitMethodSignature, 
+request&response DataModel names, and the remote RPC name. The `RemoteMethodRpc` dataclass contains 
+all the runtime-specific assumptions  and is the main point where overlap happens. The 
+`RemoteModelFinder`, which constructs `RemoteMethodRpc`s, should not import `caikit.runtime`; 
+however, it will re-use a lot of code around the different naming schemes.
+
+We created the following pseudocode to help illustrate how the `RemoteModelFinder` constructs a 
+`RemoteModuleRpc`. (Note this is the function used in the pseudocode for the [RemoteModelFinder](#remotemodelfinder))
+
+```python
+def generate_config_for_module(module: ModuleBase, connection_info: Dict[str, Any], model_path: str)->RemoteModuleConfig:
+    # Gather a description of all tasks and their associated methods
+    task_methods = []
+    for task_class in module.tasks:
+        task_functions = []
+        for input, output, signature in module_class.get_inference_signatures(task_class):
+            # Construct request_dm_name,  task_request_name, and rpc_name. This makes assumptions
+            # about caikit.runtime service generation
+            request_dm_name ~= "TaskRequest"
+            task_request_name ~= "TaskPredict"
+            rpc_name = "/api/v1/TaskPredict"
```

Comment:
Can we avoid the coupling-by-naming-scheme by having the remote model finder lookup this information as well?

We could add the supported tasks and relevant names to the new `ModelInfoResponse`, and/or we could return the supported rpc names and use the grpc reflection api to go look up the names and schemas of the request and response.

Reply:
🌶️ , yeah that's still probably the easiest way to get this done. If these schemes changed then it would be a headache, but that's just a "what if"